internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -0,0 +1,585 @@
+	// to the observed state.
+	xrConnDetails, err := c.composite.FetchConnection(ctx, xr)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errFetchXRConnectionDetails)
+	}
+	o, err := AsState(xr, xrConnDetails, observed)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errBuildObserved)
+	}
+	d, err := AsState(xr, xrConnDetails, ComposedResourceStates{})
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errBuildDesired)
+	}
+
+	events := []event.Event{}
+
+	// Run any Composition Functions in the pipeline. Each Function may mutate
+	// the desired state returned by the last, and each Function may produce
+	// results that will be emitted as events.
+	for _, fn := range req.Revision.Spec.Pipeline {
+		req := &v1beta1.RunFunctionRequest{Observed: o, Desired: d}
+
+		if fn.Input != nil {
+			in := &structpb.Struct{}
+			if err := in.UnmarshalJSON(fn.Input.Raw); err != nil {
+				return CompositionResult{}, errors.Wrapf(err, errFmtUnmarshalPipelineStepInput, fn.Step)
+			}
+			req.Input = in
+		}
+
+		// TODO(negz): Generate a content-addressable tag for this request.
+		// Perhaps using https://github.com/cerbos/protoc-gen-go-hashpb ?
+		rsp, err := c.pipeline.RunFunction(ctx, fn.FunctionRef.Name, req)
+		if err != nil {
+			return CompositionResult{}, errors.Wrapf(err, errFmtRunPipelineStep, fn.Step)
+		}
+
+		d = rsp.GetDesired()
+
+		// Results of fatal severity stop the Composition process. Other results
+		// are accumulated to be emitted as events by the Reconciler.
+		for _, rs := range rsp.Results {
+			switch rs.Severity {
+			case v1beta1.Severity_SEVERITY_FATAL:
+				return CompositionResult{}, errors.Errorf(errFmtFatalResult, fn.Step, rs.Message)
+			case v1beta1.Severity_SEVERITY_WARNING:
+				events = append(events, event.Warning(reasonCompose, errors.Errorf("Pipeline step %q: %s", fn.Step, rs.Message)))
+			case v1beta1.Severity_SEVERITY_NORMAL:
+				events = append(events, event.Normal(reasonCompose, fmt.Sprintf("Pipeline step %q: %s", fn.Step, rs.Message)))
+			case v1beta1.Severity_SEVERITY_UNSPECIFIED:
+				// We could hit this case if a Function was built against a newer
+				// protobuf than this build of Crossplane, and the new protobuf
+				// introduced a severity that we don't know about.
+				events = append(events, event.Warning(reasonCompose, errors.Errorf("Pipeline step %q returned a result of unknown severity (assuming warning): %s", fn.Step, rs.Message)))
+			}
+		}
+	}
+
+	// TODO(negz): Both here and in the PTComposer we're relying on the fact
+	// that the calling Reconciler will update xr's status in order to persist
+	// any changes we make to it. It would be more obvious if we owned making
+	// our own changes.
+
+	// TODO(negz): Because we create, modify, and apply a new XR here we'll
+	// never apply any changes to status. I think this is what is causing
+	// https://github.com/crossplane/crossplane/issues/3812.
```

Comment:
Should a function be able to modify XR spec? Or should they only change the status? 
In the PTComposer, we allow modification of XR spec, however, this is against what is proposed in this ticket: https://github.com/crossplane/crossplane/issues/4581 and causing various problems regarding what to do next with that modified XR spec, e.g. Should it be propagated up to Claim? What if the Claim has the same spec with another value? 

I think this PR is a good chance to build to correct experience here which I believe is a function that should only be able to modify the status of an XR and we should explicitly return an error if done otherwise.

Reply:
I ended up adding a commit that addresses #4581 - or at least sets us up to address it when Functions are in use.